drivers/flash/flash_andes_qspi.c

```diff
@@ -0,0 +1,950 @@
+
+
+/**
+ * @brief Write the status register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param sr The new value of the status register
+ *
+ * @return 0 on success or a negative error code.
+ */
+static int flash_andes_qspi_wrsr(const struct device *dev,
+			uint8_t sr)
+{
+	int ret = flash_andes_qspi_cmd_write(dev, FLASH_ANDES_CMD_WREN);
+
+	if (ret == 0) {
+		ret = flash_andes_qspi_access(dev, FLASH_ANDES_CMD_WRSR,
+					ANDES_ACCESS_WRITE, 0, &sr,
+					sizeof(sr));
+		flash_andes_qspi_wait_until_ready(dev);
+	}
+
+	return ret;
+}
+
+static int flash_andes_qspi_read(const struct device *dev,
+		off_t addr, void *dest, size_t size)
+{
+	const size_t flash_size = dev_flash_size(dev);
+	int ret;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > flash_size)) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	ret = flash_andes_qspi_cmd_addr_read(dev,
+		FLASH_ANDES_CMD_4READ, addr, dest, size);
+
+	release_device(dev);
+	return ret;
+}
+
+static int flash_andes_qspi_write(const struct device *dev, off_t addr,
+			 const void *src,
+			 size_t size)
+{
+	const size_t flash_size = dev_flash_size(dev);
+	const uint16_t page_size = dev_page_size(dev);
+	int ret = 0;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > flash_size)) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	ret = flash_andes_qspi_write_protection_set(dev, false);
+
+	if (ret == 0) {
+		while (size > 0) {
+			size_t to_write = size;
+
+			/* Don't write more than a page. */
+			if (to_write >= page_size) {
+				to_write = page_size;
+			}
+
+			/* Don't write across a page boundary */
+			if (((addr + to_write - 1U) / page_size) !=
+			(addr / page_size)) {
+				to_write = page_size - (addr % page_size);
+			}
+
+			flash_andes_qspi_cmd_write(dev, FLASH_ANDES_CMD_WREN);
+
+			ret = flash_andes_qspi_cmd_addr_write(dev,
+				FLASH_ANDES_CMD_4PP, addr, src, to_write);
+
+			if (ret != 0) {
+				goto out;
+			}
+
+			size -= to_write;
+			src = (const uint8_t *)src + to_write;
+			addr += to_write;
+
+			flash_andes_qspi_wait_until_ready(dev);
+		}
+	}
+
+	int ret2 = flash_andes_qspi_write_protection_set(dev, true);
+
+	if (!ret) {
+		ret = ret2;
+	}
+
+out:
+	release_device(dev);
+	return ret;
+}
+
+static int flash_andes_qspi_erase(const struct device *dev,
+		off_t addr, size_t size)
+{
+	const size_t flash_size = dev_flash_size(dev);
+	int ret = 0;
+
+	/* erase area must be subregion of device */
+	if ((addr < 0) || (size == 0) || ((size + addr) > flash_size)) {
+		return -EINVAL;
+	}
+
+	/* address must be sector-aligned */
+	if (!SPI_NOR_IS_SECTOR_ALIGNED(addr)) {
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sectors */
+	if ((size % SPI_NOR_SECTOR_SIZE) != 0) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	ret = flash_andes_qspi_write_protection_set(dev, false);
+
+	while ((size > 0) && (ret == 0)) {
```

Comment:
`ret` check  here is waste of clocks as it will be checked every loop run and it can only, except for entering loop in the first place, be non-0 in case of failure, which is in single point within the loop.
Instead of the `ret == 0` check here, the loop should be skipped over of 404 fails and else at 435 should break out of the loop.

Reply:
Thanks, I've removed `ret` checking and add `break` statement.